kernel: KProveOptions: --concrete-rules is now implemented in front-end

denis-bogdanas opened the pull request:
Option `--concrete-rules` is now implemented in the front-end and is available for any backend. Right now only `KProveOptions` has it but it can easily be added to other proof-related tools like CBMC.

ktest.mak changes: extended to allow definition .k files in other directory than current running directory. This is required for this PR, because I want to run same compiled definition with different sets of kprove options, to test correct caching/non-caching.

denis-bogdanas committed:
Replaced "concrete" string with Attribute.CONCRETE_KEY where possible.

denis-bogdanas committed:
KProveOptions: --concrete-rules is now implemented in front-end.

dwightguth commented on haskell-backend/src/main/java/org/kframework/backend/haskell/HaskellBackend.java:
```diff
@@ -20,9 +22,10 @@
     @Inject
     public HaskellBackend(
             KompileOptions kompileOptions,
+            KProveOptions kproveOptions,
```
Architecturally, this is wrong. It should not require a KProveOptions object to construct something unless you are within the kprove tool's code path, but backends can also be constructed by kompile

denis-bogdanas replied:
Yes I'm aware of this. Actually there are other changes needed that will make access to KProveOptions unnecessary.

denis-bogdanas committed:
tests/concrete-function-cache - test exposing interaction between incorrect implementation of --concrete-rules in this PR and runtime definition cache (kserver mode).
ktest.mak: extension to allow $(DEF) in different directory. Required for this test.

denis-bogdanas committed:
KProveOptions: --concrete-rules is now implemented in front-end. Entirely new implementation that doesn't depend on KProveOptions.

denis-bogdanas committed:
optimization

denis-bogdanas committed:
Merge branch 'master' into PR-concrete-rules

denis-bogdanas committed:
Merge branch 'master' into PR-concrete-rules
